lib/opam_solve.ml

```diff
@@ -167,8 +189,13 @@ module Opam_monorepo_context (Base_context : BASE_CONTEXT) :
         in
         [ (version, Ok opam_file) ]
     | None ->
-        Base_context.candidates base_context name
-        |> filter_candidates ~allow_jbuilder ~require_dune ~name
+        let candidates = Base_context.candidates base_context name in
+        let must_cross_compile =
+          prefer_cross_compile
+          && List.exists ~f:candidate_cross_compile candidates
```

Comment:
It has to work this way because no package has the tag by default. What problem are you foreseeing here exactly?

The explicit user preference, is to select versions tagged with `cross-compile` if any exists, as it is documented.

Reply:
My concern is that if the candidates don't have any that is tagged that this preference is discarded silently and a version is selected that does not cross compile. I would suggest to notify the user if their preference doesn't do anything (via a warning or something), so if they pick the option but they get the warning they know something is not right, instead of silently doing a thing successfully (locking non-`cross-compile` versions because a repo wasn't added) that the user didn't intend.

Reply:
I started implementing this and while doing so I realized that it's not ideal. There are perfectly valid reasons why one would set the flag but it would still have no effect. For instance, at the moment, if you want to cross compile but don't depend on zarith, the flag will have no effect, no matter how you set things up. You still might want to enable the flag in case things change in the future.

We don't want to scan the entire space to see if any package has the tag set. It can be a costly operation and it's just for a warning here. This approach would also not be very accurate as if we ever upstream that tag somewhere, the warning would never be triggered. We're not there yet but it's worth considering it while we design this.

That puts us in the situation where we have to choose between not warning users in cases where they forgot to set the right repos or wrongly warning users that they might have forgot about something when they did not.

Here I would be in favor of not adding this warning (for now!). The only known consumer atm is mirage. In that case the tool is responsible for setting the mirage repo so this should not occur. It's extra code to maintain, especially ugly one since it's not easy to add without some mutable state. Once we have more users of this feature, it will become clearer whether this is worth doing.

Reply:
I don't thing adding flags "for future use" is a good pattern and reality shows that users only add options when stuff breaks ;-)

But I agree that it could be expensive. Let's keep it out and see if it causes actual real-life problems in the future that an user would complain about.